pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,391 @@
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, dirDelim) + dirDelim
+	}
+
+	for object := range b.ListObjects(ctx, dir, false) {
+		if object.err != nil {
+			return object.err
+		}
+		if object.key == "" {
+			continue
+		}
+		if err := f(object.key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if len(name) == 0 {
+		return nil, errors.Errorf("given object name should not empty")
+	}
+
+	opts := &cos.ObjectGetOptions{}
+	if length != -1 {
+		if err := setRange(opts, off, off+length-1); err != nil {
+			return nil, err
+		}
+	}
+
+	resp, err := b.client.Object.Get(ctx, name, opts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := resp.Body.Read(nil); err != nil {
+		runutil.CloseWithLogOnErr(b.logger, resp.Body, "cos get range obj close")
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	if _, err := b.client.Object.Head(ctx, name, nil); err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "head cos object")
+	}
+
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	tmpErr := toErrorResponse(err)
+	if tmpErr.Code == "NoSuchKey" {
+		return true
+	}
+	if tmpErr.Response != nil && tmpErr.Response.StatusCode == http.StatusNotFound {
+		return true
+	}
+	return false
+}
+
+func toErrorResponse(err error) *cos.ErrorResponse {
+	switch err := err.(type) {
+	case *cos.ErrorResponse:
+		return err
+	default:
+		return &cos.ErrorResponse{}
+	}
+}
+
+func (b *Bucket) Close() error { return nil }
+
+type objectInfo struct {
+	key string
+	err error
+}
+
+func (b *Bucket) ListObjects(ctx context.Context, objectPrefix string, recursive bool) <-chan objectInfo {
+	objectsCh := make(chan objectInfo, 1)
+	delimiter := "/"
+	if recursive {
+		delimiter = ""
+	}
+
+	go func(objectsCh chan<- objectInfo) {
+		defer close(objectsCh)
+		var marker string
+		for {
+			result, _, err := b.client.Bucket.Get(ctx, &cos.BucketGetOptions{
+				Prefix:    objectPrefix,
+				MaxKeys:   1000,
+				Marker:    marker,
+				Delimiter: delimiter,
+			})
+			if err != nil {
```

Comment:
@bwplotka  I acknowledge your concern, but I think the `channel` was a buffered channel, it will not be blocked at the first time on writing and the goroutine and channel will be closed whatever the consumer received the error.  Is that wrong for my opinion? 

But I should update the code to wait the consumer received the error and then come out the `f`. Thank you for your suggestion.